src/Task.cc

```diff
@@ -303,23 +303,44 @@ static vector<uint8_t> ptrace_get_regs_set(Task* t, const Registers& regs,
 }
 
 template <typename Arch>
-static int64_t get_io_offset_arch(int syscallno, const Registers& regs) {
+static int64_t get_io_offset_arch(Task* t, int syscallno,
+                                  const Registers& regs) {
   switch (syscallno) {
     case Arch::pwrite64:
     case Arch::pwritev:
     case Arch::pread64:
-    case Arch::preadv:
+    case Arch::preadv: {
       if (sizeof(typename Arch::unsigned_word) == 4) {
         return regs.arg4() | (uint64_t(regs.arg5_signed()) << 32);
       }
       return regs.arg4_signed();
+    }
+    case Arch::writev:
+    case Arch::write: {
+      ASSERT(t, !t->session().is_replaying())
+          << "Can only read a file descriptor's offset while recording";
+      int fd = regs.arg1_signed();
+      // Get the offset from /proc/*/fdinfo/*
+      char fdinfo_path[PATH_MAX];
+      sprintf(fdinfo_path, "/proc/%d/fdinfo/%d", t->tid, fd);
+      FILE* fdinfo_file;
+      if (!(fdinfo_file = fopen(fdinfo_path, "r")))
+        FATAL() << "Failed to open " << fdinfo_path;
+      int64_t offset = -1;
+      if (fscanf(fdinfo_file, "pos:\t%ld", &offset) != 1)
+        FATAL() << "Failed to read position";
+      fclose(fdinfo_file);
+      // The pos we just read, was after the write completed. Luckily, we do
+      // know how many bytes were written.
```

Comment:
There's a potential issue here for concurrent writes to the same file.

I *think* we should probably make the write `PREVENT_SWITCH` if we're writing to a mapped file. That should fix it.